tests/foreman/longrun/test_oscap.py

```diff
@@ -42,670 +42,593 @@
-        end-to-end tests like OSCAP etc
-
-        :returns: A dict of entities to help with provisioning
-        """
-        cls.rhel6_content = OSCAP_DEFAULT_CONTENT['rhel6_content']
-        cls.rhel7_content = OSCAP_DEFAULT_CONTENT['rhel7_content']
-        cls.rhel8_content = OSCAP_DEFAULT_CONTENT['rhel8_content']
-        sat6_hostname = settings.server.hostname
-        proxy = Proxy.list({'search': sat6_hostname})[0]
-        p_features = set(proxy.get('features').split(', '))
-        if {'Puppet', 'Ansible', 'Openscap'}.issubset(p_features):
-            cls.proxy_id = proxy.get('id')
-        else:
-            raise ProxyError('Some features like Puppet, DHCP, Openscap, Ansible are not present')
-        ak_name_8 = gen_string('alpha')
-        ak_name_7 = gen_string('alpha')
-        ak_name_6 = gen_string('alpha')
-        repo_values = [
-            {'repo': settings.sattools_repo['rhel8'], 'akname': ak_name_8},
-            {'repo': settings.sattools_repo['rhel7'], 'akname': ak_name_7},
-            {'repo': settings.sattools_repo['rhel6'], 'akname': ak_name_6},
-        ]
-        # Create new organization and environment.
-        org = entities.Organization(name=gen_string('alpha')).create()
-        cls.puppet_env = (
-            entities.Environment().search(query={'search': 'name=production'})[0].read()
-        )
-        cls.puppet_env.organization.append(org)
-        cls.puppet_env = cls.puppet_env.update(['organization'])
-        smart_proxy = (
-            entities.SmartProxy().search(query={'search': f'name={sat6_hostname}'})[0].read()
+def fetch_scap_and_profile_id(scap_name, scap_profile):
+    """Extracts the scap ID and scap profile id
+
+    :param scap_name: Scap title
+    :param scap_profile: Scap profile you want to select
+
+    :returns: scap_id and scap_profile_id
+    """
+
+    default_content = Scapcontent.info({'title': scap_name}, output_format='json')
+    scap_id = default_content['id']
+    scap_profile_ids = [
+        profile['id']
+        for profile in default_content['scap-content-profiles']
+        if scap_profile in profile['title']
+    ]
+    return scap_id, scap_profile_ids
+
+
+@pytest.fixture(scope='module')
+def scap_setup(request, module_org):
+    """Sets up the whole provisioning environment needed for Puppet based
+    end-to-end tests like OSCAP etc
+
+    :returns: A dict of entities to help with provisioning
+    """
+    request.puppet_classes = ['foreman_scap_client::params', 'foreman_scap_client']
+    rhel6_content = OSCAP_DEFAULT_CONTENT['rhel6_content']
+    rhel7_content = OSCAP_DEFAULT_CONTENT['rhel7_content']
+    rhel8_content = OSCAP_DEFAULT_CONTENT['rhel8_content']
+    sat6_hostname = settings.server.hostname
+    proxy = Proxy.list({'search': sat6_hostname})[0]
+    p_features = set(proxy.get('features').split(', '))
+    if {'Puppet', 'Ansible', 'Openscap'}.issubset(p_features):
+        request.proxy_id = proxy.get('id')
```

Comment:
I don't think this is a pattern that should be used. the `request` fixture in pytest is special, and is not a generic object that we should be attaching attributes to.

You will see some use of setting `request.cls` attributes, which is different. In this case there is no longer a class.

The pattern that you should follow here is defining separate fixtures for the entities that are created during setup, and are utilized by the test.

There should be a `proxy_module` fixture defined, which actually _creates_ a proxy to use, not just assumes that there is already one configured.  It should then return that proxy entity, or its ID if that is all that is needed.

There should also be fixtures defined for the use of `request.ak_name`, `request.org`, `request.puppet_env`, `request.env`, and all others below.

Please see this PR as an example of how to scope and structure the fixtures:
https://github.com/SatelliteQE/robottelo/pull/8441/files

The test cases then include what fixtures they need to directly access, and use them directly, instead of this generic `setup` fixture returning the request instance which has been mangled by the fixture itself.

Reply:
Another option for a pattern to follow here if you would really prefer to have one setup fixture is to define a `namedtuple` that can be used to return all of the relevant entities through one fixture return.

```
from collections import namedtuple
SCAPEntities = namedtuple('SCAPEntities', ['proxy', 'ak_name', 'org'])  # short example, include all fields

@pytest.fixture
def scap_setup(module_org):
    proxy=entities.list()[0]
    ak_name={'rhel': 'random-string'}
    return SCAPEntities(proxy, ak_name, module_org)
   
def test_something(scap_setup):
    logger.info(f'The Proxy used for scap testing is: {scap_setup.proxy}')
    # reference scap_setup.<field> for all of the things attached to the request instance as currently returned from scap_setup   
```

Reply:
updated pr